### PR TITLE
Move to Forge

### PR DESCRIPTION
**Pull request by rockyfour**
This is needed for https://github.com/makerdao/exchange-callees/pull/34.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 /out
+/cache
```
Comment (naszam) [inline .gitignore:2]: missing eol

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
-	hevm flatten --source-file "src/UniswapV2Callee.sol" > out/UniswapV2CalleeDai.sol
-	hevm flatten --source-file "src/UniswapV2LpTokenCallee.sol" > out/UniswapV2LpTokenCalleeDai.sol
-	hevm flatten --source-file "src/UniswapV3Callee.sol" > out/UniswapV3Callee.sol
-	hevm flatten --source-file "src/WstETHCurveUniv3Callee.sol" > out/WstETHCurveUniv3Callee.sol
-	hevm flatten --source-file "src/CurveLpTokenUniv3Callee.sol" > out/CurveLpTokenUniv3Callee.sol
+	forge flatten "src/UniswapV2Callee.sol" > out/UniswapV2CalleeDai.sol
+	forge flatten "src/UniswapV2LpTokenCallee.sol" > out/UniswapV2LpTokenCalleeDai.sol
+	forge flatten "src/UniswapV3Callee.sol" > out/UniswapV3Callee.sol
+	forge flatten "src/WstETHCurveUniv3Callee.sol" > out/WstETHCurveUniv3Callee.sol
+	forge flatten "src/CurveLpTokenUniv3Callee.sol" > out/CurveLpTokenUniv3Callee.sol
```
Comment (naszam) [inline Makefile:13]: does it works the same? I think there were some open issues with `forge flatten`

Comment (naszam) [reply Makefile:13]: tested now, and with the path `out/flatten.sol` doesn't seem to work

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -1,25 +1,20 @@
...
+        with:
+          version: nightly
 
-    - name: Use maker and dapp cachix
-      uses: cachix/cachix-action@v10
-      with:
-        name: maker
-        extraPullNames: dapp
+      - name: Install Dependencies
+        run: forge install
```
Comment (naszam) [inline .github/workflows/tests.yaml:15]:
```suggestion
        run: git submodule update --init --recursive
```
I think this way was preferred to install the dependencies

File: README.md [hunk 4]
```diff
@@ -38,10 +38,10 @@ To slow down or defend against [generalized frontrunning bots](https://medium.co
 - Using a mempool-shielded transaction service, as outlined in [this blog post](https://samczsun.com/escaping-the-dark-forest/)
 
 ## Testing
-Requires [Dapptools](https://github.com/dapphub/dapptools)
+Requires [Forge](https://github.com/foundry-rs/foundry)
 ```
-$ dapp update
-$ dapp --use solc:0.6.12 test
+$ forge install
+$ forge test --use solc:0.6.12
```
Comment (naszam) [inline README.md:44]:
```suggestion
$ make test
```
maybe the script can be preferred here

File: test.sh [hunk 5]
```diff
@@ -1,7 +1,8 @@
 #! /bin/bash
```
Comment (naszam) [inline test.sh:1]: nit: 
```suggestion
#!/usr/bin/env bash
```

File: Makefile (hunk 2 above)
Comment (rockyfour) [reply Makefile:13]: The diff seems ok - https://www.diffchecker.com/ilFDi7qj/
Was also able to verify the contract using forge flatten.
So I think we are good.

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (rockyfour) [reply .github/workflows/tests.yaml:15]: I copied the `forge intstall` commad from the D3M repo yaml.
Do you know why is it not good?

File: test.sh (hunk 5 above)
Comment (rockyfour) [reply test.sh:1]: Can you please explain this suggestion?

Comment (naszam) [reply test.sh:1]: It's a common pattern used in other scripts (see spell repos), I specified it as a nit as not related to the PR changes but just for consistency.
The rational behind the shift in pattern I guess is related to giving more flexibility on different systems.

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (naszam) [reply .github/workflows/tests.yaml:15]: It should install missing dependencies automatically when testing, so it should not even be needed here, but I guess if we want to specified it the `git` way helps fixing some edge cases I guess.

File: Makefile (hunk 2 above)
Comment (naszam) [reply Makefile:13]: see: https://github.com/foundry-rs/foundry/issues/4034

Comment (naszam) [reply Makefile:13]: not sure if we're safe here but worth double checking

Comment (naszam) [reply Makefile:13]: also, with forge to specify output path it seems to be preferred the use of the option `--output` (see [here](https://book.getfoundry.sh/reference/forge/forge-flatten?highlight=forge%20flatten#examples))

File: scripts/deploy-mainnet.sh [hunk 6]
```diff
@@ -1,32 +1,32 @@
...
-  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2)
+  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2
 
-CurveLpTokenUniv3Callee=$(dapp create CurveLpTokenUniv3Callee \
+echo "CurveLpTokenUniv3Callee:"
+forge create CurveLpTokenUniv3Callee.sol:CurveLpTokenUniv3Callee --constructor-args \
   0xE592427A0AEce92De3Edee1F18E0157C05861564 \
   0x9759A6Ac90977b93B58547b4A71c78317f391A28 \
-  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2)
+  0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2
```
Comment (godsflaw) [inline scripts/deploy-mainnet.sh:29]: I feel like the output is less clean like this, but I used to publish these back when we had a keeper coordination channel.  Since that doesn't exist anymore, I'll just register my protest, but not hold this PR up because of it.

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (godsflaw) [reply .github/workflows/tests.yaml:15]: I don't know about this "git submodule update --init --recursive" issue @naszam is bringing up.  I know there was some problems with "forge install" recursing at one point, and it probably still has that issue, but if we can get the repo working with "forge install", I think we should strive for that.

File: test.sh (hunk 5 above)
Comment (godsflaw) [reply test.sh:1]: Yes, this increases portability.  I support this change.

File: Makefile [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
 deploy-mainnet  :; make && ./scripts/deploy-mainnet.sh
 flatten         :;
-	hevm flatten --source-file "src/UniswapV2Callee.sol" > out/UniswapV2CalleeDai.sol
-	hevm flatten --source-file "src/UniswapV2LpTokenCallee.sol" > out/UniswapV2LpTokenCalleeDai.sol
-	hevm flatten --source-file "src/UniswapV3Callee.sol" > out/UniswapV3Callee.sol
-	hevm flatten --source-file "src/WstETHCurveUniv3Callee.sol" > out/WstETHCurveUniv3Callee.sol
-	hevm flatten --source-file "src/CurveLpTokenUniv3Callee.sol" > out/CurveLpTokenUniv3Callee.sol
+	forge flatten "src/UniswapV2Callee.sol" > out/UniswapV2CalleeDai.sol
+	forge flatten "src/UniswapV2LpTokenCallee.sol" > out/UniswapV2LpTokenCalleeDai.sol
+	forge flatten "src/UniswapV3Callee.sol" > out/UniswapV3Callee.sol
```
Comment (godsflaw) [inline Makefile:11]: We need to change the output file name, or directory to `out/flatten` to prevent this conflict:

```bash
$ rm -fr out
$ make clean
forge clean
$ forge install
$ make
FOUNDRY_OPTIMIZER=true FOUNDRY_OPTIMIZER_RUNS=200 forge build --use solc:0.6.12
[⠊] Compiling...
[⠑] Compiling 86 files with 0.6.12
[⠰] Solc 0.6.12 finished in 5.84s
[SNIP ...]


$ make flatten
forge flatten "src/UniswapV2Callee.sol" > out/UniswapV2CalleeDai.sol
forge flatten "src/UniswapV2LpTokenCallee.sol" > out/UniswapV2LpTokenCalleeDai.sol
forge flatten "src/UniswapV3Callee.sol" > out/UniswapV3Callee.sol
/bin/sh: out/UniswapV3Callee.sol: Is a directory
make: *** [flatten] Error 1
```

**Review by godsflaw — CHANGES_REQUESTED** [review]

File: scripts/deploy-mainnet.sh (hunk 6 above)
Comment (godsflaw) [reply scripts/deploy-mainnet.sh:29]: Looks like we just need to change the `mainnet` ones to: `src/<file>.sol:<file>Dai`.  Where the mainnet ones are missing `src/`.

File: .github/workflows/tests.yaml [hunk 8]
```diff
@@ -1,25 +1,20 @@
-on: [push, pull_request]
+on: [push, pull request]
```
Comment (naszam) [inline .github/workflows/tests.yaml:1]: CI doesn't seem to work in the PR commit, have the secret been set up already?
Also, I would make sure that the CI job is correct regarding indentations and so on.

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (naszam) [reply .github/workflows/tests.yaml:15]: maybe adopting a `foundry.toml` could ensure a more compact config settings and eventually address those issues if they arise.

File: .gitignore (hunk 1 above)
Comment (rockyfour) [reply .gitignore]: [2667bd7](https://github.com/makerdao/exchange-callees/pull/36/commits/2667bd785d7e69c3046e40cb1fbc743c6fc5f90d)

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (naszam) [reply .github/workflows/tests.yaml:15]: but yeah, let's get the CI running and if it passes we can go with that for now

File: .github/workflows/tests.yaml [hunk 9]
```diff
@@ -1,25 +1,20 @@
...
 
 jobs:
   tests:
     runs-on: ubuntu-latest
     steps:
-    - name: Checkout repository and submodules
-      uses: actions/checkout@v2
-      with:
-        submodules: recursive
+      - uses: actions/checkout@v3
```
Comment (naszam) [inline .github/workflows/tests.yaml:7]: It seems it's not aligned correctly, this could explain why it's not running in the PR, also it's missing the `name` in the steps for checkout.

File: Makefile (hunk 2 above)
Comment (naszam) [reply Makefile:13]: Addressed in #37

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (naszam) [reply .github/workflows/tests.yaml:15]: It seems working in #37

File: README.md (hunk 4 above)
Comment (naszam) [reply README.md:44]: Addressed in #37

File: test.sh (hunk 5 above)
Comment (naszam) [reply test.sh:1]: Addressed in #37

File: .github/workflows/tests.yaml (hunk 8 above)
Comment (naszam) [reply .github/workflows/tests.yaml:1]: Addressed in #37 (was missing the underscore in `pull_request`)

File: .github/workflows/tests.yaml (hunk 9 above)
Comment (naszam) [reply .github/workflows/tests.yaml:7]: Addressed in #37

File: Makefile (hunk 7 above)
Comment (naszam) [reply Makefile]: Addressed in #37

**Review by naszam — APPROVED** [review]
LGTM (Pending secret update before merging)
